src/frontend/src/handler/alter_table_column.rs

```diff
@@ -97,6 +118,64 @@ pub async fn replace_table_with_definition(
     Ok(())
 }
 
+pub(crate) fn hijack_merger_for_target_table(
+    graph: &mut StreamFragmentGraph,
+    target_columns: &[ColumnCatalog],
+    sink: &SinkCatalog,
+) -> Result<()> {
+    let mut sink_columns = sink.original_target_columns.clone();
+    if sink_columns.is_empty() {
+        // This is due to the fact that the value did not exist in earlier versions,
+        // which means no schema changes such as `ADD/DROP COLUMN` have been made to the table.
+        // Therefore the columns of the table at this point are `original_target_columns`.
+        // This value of sink will be filled on the meta.
+        sink_columns = target_columns.to_vec();
+    }
+
+    let mut i = 0;
+    let mut j = 0;
+    let mut exprs = Vec::new();
+
+    while j < target_columns.len() {
+        if i < sink_columns.len() && sink_columns[i].data_type() == target_columns[j].data_type() {
+            exprs.push(ExprImpl::InputRef(Box::new(InputRef {
+                data_type: sink_columns[i].data_type().clone(),
+                index: i,
+            })));
+
+            i += 1;
+            j += 1;
+        } else {
+            exprs.push(ExprImpl::Literal(Box::new(Literal::new(
+                None,
+                target_columns[j].data_type().clone(),
+            ))));
+
+            j += 1;
+        }
+    }
+
+    let pb_project = PbNodeBody::Project(ProjectNode {
+        select_list: exprs.iter().map(|expr| expr.to_expr_proto()).collect(),
+        ..Default::default()
+    });
+
+    for fragment in graph.fragments.values_mut() {
+        if let Some(node) = &mut fragment.node {
+            insert_merger_to_union_with_project(
+                node,
+                &pb_project,
+                &format!(
```

Comment:
I speculate that you are referring to two issues here?
The first point is that a project operator is manually constructed here.
The second point is that a unique name is encoded to recognize in the meta which project identifies which upstream.
I believe it would be appropriate to add a new stream plan node (leaf node) in the optimizer. However, that would be a huge change. I think it's okay for now, and we can complete it in a future PR